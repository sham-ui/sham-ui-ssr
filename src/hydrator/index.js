import Storage from './storage/service';
import ServerSideHooks from './hooks/service';
import ServerSideDOM from './dom/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Main function for setup hydrating components
 * @param {Object} DI App DI container
 */
export function setup( DI ) {
    const storage = new Storage();

    // Override default hooks service
    DI.bind(
        HOOKS_DI_KEY,
        new ServerSideHooks( storage )
    );

    // Override default dom service
    DI.bind(
        DOM_DI_KEY,
        new ServerSideDOM( storage )
    );

    return storage.root;
}


/**
 * Hydrate all components
 * @param {Object} DI App DI container
 * @return {Promise}
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import { setupHydrator, hydrate } from 'sham-ui-ssr/hydrator';
 *
 * const DI = createDI();
 * const root = setupHydrator( DI );
 * mainInitializer( DI, root );
 * start( DI );
 * hydrate( DI ).then( storage => {
 *     const { html, data } = storage.hydrate();
 * } )
 */
export function hydrate( DI ) {
    return DI.resolve( HOOKS_DI_KEY ).hydrating();
}
