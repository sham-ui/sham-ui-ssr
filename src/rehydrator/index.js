import RestoreDom from './dom/service';
import RehydrateHooks from './hooks/service';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

/**
 * Save hooks & dom services for restore after rehydrating
 * @param {Object} DI App DI container
 * @return {Function}
 */
function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

/**
 * Main function for rehydrating
 * @param {Object} DI App DI container
 * @param {Object} storage
 * @return {Function} Restore function for disable rehydrating and switch to default behavior
 *
 * @example
 * import { createDI, start } from 'sham-ui';
 * import setupUnsafe from 'sham-ui-unsafe';
 * import { setupRehydrator } from 'sham-ui-ssr/rehydrator';
 *
 * const DI = createDI();
 * const data = JSON.parse( document.getElementById( 'data' ).innerText );
 * const disableRehydrating = setupRehydrator( DI, data );
 * mainInitializer( DI, root );
 * start( DI );
 * disableRehydrating();
 * setupUnsafe( DI );
 */
export function setup( DI, storage ) {
    const restore = makeRestore( DI );
    DI.bind( HOOKS_DI_KEY, new RehydrateHooks( storage ) );
    DI.bind( DOM_DI_KEY, new RestoreDom( storage ) );
    return restore;
}
