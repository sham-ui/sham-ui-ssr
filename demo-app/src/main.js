import './styles/main.scss';
import setupUnsafe from 'sham-ui-unsafe';
import { createDI, start } from 'sham-ui';
import { setup as setupHydrator, hydrate } from '../../src/hydrator/index';
import { setup as setupRehydrator } from '../../src/rehydrator/index';
import mainInitializer from './initializers/main';

const HOOKS_DI_KEY = 'sham-ui:hooks';
const DOM_DI_KEY = 'sham-ui:dom';

function makeRestore( DI ) {
    const originalHooks = DI.resolve( HOOKS_DI_KEY );
    const originalDom = DI.resolve( DOM_DI_KEY );
    return function() {
        DI.bind( HOOKS_DI_KEY, originalHooks );
        DI.bind( DOM_DI_KEY, originalDom );
    };
}

const DI = createDI();
const restoreOriginal = makeRestore( DI );

const root = setupHydrator( DI );
mainInitializer( DI, root );
start( DI );
hydrate( DI ).then(
    ( storage ) => {
        const getById = document.getElementById.bind( document );
        // eslint-disable-next-line no-console
        console.dir( storage );
        try {
            const { html, data } = storage.hydrate();
            getById( 'ssr-rendered' ).innerHTML = html;
            getById( 'raw-html' ).innerText = html;
            getById( 'data' ).innerText = JSON.stringify( JSON.parse( data ), null, 4 );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
        try {
            DI.resolve( 'sham-ui:store' ).byId.clear();
            const root = getById( 'client' );
            root.innerHTML = getById( 'ssr-rendered' ).innerHTML;
            restoreOriginal();
            const data = JSON.parse( document.getElementById( 'data' ).innerText );
            const disableRehydrating = setupRehydrator( DI, data );
            mainInitializer( DI, root );
            start( DI );
            disableRehydrating();
            setupUnsafe( DI );
            // eslint-disable-next-line no-console
            console.dir( DI.resolve( 'sham-ui:store' ).byId.keys() );
        } catch ( e ) {
            // eslint-disable-next-line no-console
            console.error( e );
            // eslint-disable-next-line no-debugger
            debugger;
        }
    }
);

