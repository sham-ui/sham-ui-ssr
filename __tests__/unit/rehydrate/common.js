import { compile } from 'sham-ui-test-helpers';
import { ssrAndRehydrate } from '../../../src/testing';

it( 'should render simple DOM', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`<div>Content</div>`
    );
    expect( meta.html ).toBe( '<div>Content</div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should insert variable in attributes', async() => {
    expect.assertions( 3 );
    const meta = await ssrAndRehydrate(
        compile`<input type="text" value="{{ value }}">`,
        {
            value: 'Value'
        }
    );
    expect( meta.component.nodes[ 0 ].value ).toEqual( 'Value' );
    expect( meta.html ).toBe( '<input type="text" value="Value">' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should properly work with text constants in text nodes', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`<p>foo {{ bar }} baz</p>`,
        {
            bar: 'bar'
        }
    );
    expect( meta.html ).toBe( '<p>foo <!--0-->bar<!--1--> baz</p>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should properly work with text constants in attributes', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`<div class="foo {{ bar }} baz"></div>`,
        {
            bar: 'bar'
        }
    );
    expect( meta.html ).toBe( '<div class="foo baz bar"></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should save value for variables in complex cases', async() => {
    expect.assertions( 4 );
    const meta = await ssrAndRehydrate(
        compile`<div class="{{ foo }} {{ bar }}"></div>`,
        {
            foo: 'first',
            bar: 'second'
        }
    );
    expect( meta.html ).toBe( '<div class="first second"></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
    meta.component.update( {
        foo: 'updated'
    } );
    expect( meta.ctx.container.innerHTML ).toBe( '<div class="updated second"></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should properly work with more then one node on topmost level', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`
        <p>first</p>

        <p>second</p>
        `
    );
    expect( meta.html ).toBe( '<p>first</p><p>second</p>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should optimize "if"/"for" tag, if it is only child', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`
        <div>
            <p>
                {% if a %}a{% endif %}
            </p>

            <p>
                {% for b %}b{% endfor %}
            </p>
        </div>
        `,
        {
            a: true,
            b: [ 1 ]
        }
    );
    expect( meta.html ).toBe( '<div><p>a</p><p>b</p></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should place placeholders for multiply "if" tags', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`
        <div>
            {% if a %}a{% endif %}
            {% if b %}b{% endif %}
        </div>
        `
    );
    expect( meta.html ).toBe( '<div><!--0--><!--1--></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should place placeholders for multiply "if" and "for" tags', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`
        <div>
            {% if a %}a{% endif %}
            {% for b %}i{% endfor %}
        </div>
        `
    );
    expect( meta.html ).toBe( '<div><!--0--><!--1--></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should properly for with filters', async() => {
    expect.assertions( 2 );
    const filters = {
        append: function( value, text ) {
            return value + text;
        },
        upperCase: function( value ) {
            return value.toUpperCase();
        }
    };
    const meta = await ssrAndRehydrate(
        compile`<p>{{ text | append('case') | upperCase }}</p>`,
        { text: 'upper_' },
        { filters }
    );
    expect( meta.html ).toBe( '<p>UPPER_CASE</p>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should work with expressions', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`
        <a title="{{ (a + b) * c / d }}">{{ more.amazing + features[0] + features[1] }}</a>
        `,
        {
            a: 1,
            b: 2,
            c: 100,
            d: 2,
            more: {
                amazing: 'a'
            },
            features: [ 'b', 'c' ]
        }
    );
    expect( meta.html ).toBe( '<a title="150">abc</a>' );
    meta.component.update( { more: { amazing: 'Amazing!' } } );
    expect( meta.ctx.container.innerHTML ).toBe( '<a title="150">Amazing!bc</a>' );
} );

it( 'should render empty attributes', async() => {
    expect.assertions( 3 );
    const meta = await ssrAndRehydrate(
        compile`<input type="checkbox" value="" checked>`
    );
    expect( meta.html ).toBe( '<input type="checkbox" value="" checked>' );
    expect( meta.component.nodes[ 0 ].checked ).toEqual( true );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should render attributes without quotes', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`<div class={{ name }}></div>`,
        {
            name: 'name'
        }
    );
    expect( meta.html ).toBe( '<div class="name"></div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should support global variables', async() => {
    expect.assertions( 4 );
    const first = await ssrAndRehydrate(
        compile`
        <i>
            {{ host == window.location.host ? 'expr' : '' }},
            {% if host == window.location.host %}if{% endif %},
            {% for i of [host == window.location.host] %}{{ i ? 'for' : '' }}{% endfor %},
            <i class="{{ host == window.location.host ? 'attr' : '' }}"></i>
        </i>
        `,
        {
            host: window.location.host
        }
    );
    expect( first.html ).toBe(
        '<i>expr<!--0-->, <!--1-->if<!--0-->, <!--2-->for<!--1-->, <i class="attr"></i></i>'
    );
    expect( first.toJSON() ).toMatchSnapshot();

    const second = await ssrAndRehydrate(
        compile`
            {{ Array.isArray(array) ? 'array' : '' }},
            {{ Math.pow(2, 2) }},
            {{ Object.keys(obj).join(';') }},
            {{ JSON.stringify(obj) }}
        `,
        {
            array: [ 1, 2, 3 ],
            obj: { a: 1, b: 2 }
        }
    );
    expect( second.html ).toBe(
        // eslint-disable-next-line max-len
        'array<!--0-->, <!--1-->4<!--2-->, <!--3-->a;b<!--4-->, <!--5-->{&quot;a&quot;:1,&quot;b&quot;:2}'
    );
    expect( second.toJSON() ).toMatchSnapshot();
} );

it( 'should support expressions without variables', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`{{ 1 + 2 * 3 }}`
    );
    expect( meta.html ).toBe( '7' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should ignore all html comments', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`
            <!-- comment with <tags> and {{ expr }}, {% tags %}, --, #, //, >. (c). -->
            <span>Moon</span>
        `
    );
    expect( meta.html ).toBe( '<span>Moon</span>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should replace HTML entities with Unicode symbols', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`&quot;&amp;&apos;&lt;&gt;&copy;&pound;&plusmn;&para;&ensp;&mdash;&emsp;&euro;&thinsp;&hearts;&notExists;`
    );
    expect( meta.html ).toBe( '&quot;&amp;&#39;&lt;&gt;©£±¶ — € ♥&amp;notExists;' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should ignore script tag content', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`<script> var result = 2 < 5;</script>`
    );
    expect( meta.html ).toBe( '<script> var result = 2 < 5;</script>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should support style attribute', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`<div \\style="background-color: red">DANGER</div>`
    );
    expect( meta.html ).toBe( '<div style="background-color: red">DANGER</div>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should support render checked == true attr', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`<input type="checkbox" checked={{true}}/>`
    );
    expect( meta.html ).toBe( '<input type="checkbox" checked>' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should support don\'t render checked == false attr', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`<input type="checkbox" checked={{false}}/>`
    );
    expect( meta.html ).toBe( '<input type="checkbox">' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should escape textContent', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`{{content}}`,
        {
            content: '<sciprt>alert(1)</script>'
        }
    );
    expect( meta.html ).toBe( '&lt;sciprt&gt;alert(1)&lt;/script&gt;' );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should escape attribute value', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`<button color={{value}}></button>`,
        {
            value: '" onclick="alert(1)"'
        }
    );
    expect( meta.html ).toBe(
        '<button color="&quot; onclick=&quot;alert(1)&quot;"></button>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );


it( 'should escape class atrribute', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`<button class={{value}}></button>`,
        {
            value: '" onclick="alert(1)"'
        }
    );
    expect( meta.html ).toBe(
        '<button class="&quot; onclick=&quot;alert(1)&quot;"></button>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'should pass real onclick atrribute', async() => {
    expect.assertions( 2 );
    const meta = await ssrAndRehydrate(
        compile`<button onclick={{value}}></button>`,
        {
            value: 'alert(1)'
        }
    );
    expect( meta.html ).toBe(
        '<button onclick="alert(1)"></button>'
    );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
