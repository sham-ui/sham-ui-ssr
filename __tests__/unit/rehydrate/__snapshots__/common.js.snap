// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should ignore all html comments 1`] = `
Object {
  "Options": Object {
    "ID": "component",
  },
  "html": "<span>Moon</span>",
}
`;

exports[`should ignore script tag content 1`] = `
Object {
  "Options": Object {
    "ID": "component",
  },
  "html": "
<script>
  var result = 2 < 5;
</script>
",
}
`;

exports[`should insert variable in attributes 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "value": "Value",
  },
  "html": "<input type=\\"text\\" value=\\"Value\\">",
}
`;

exports[`should optimize "if"/"for" tag, if it is only child 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "a": true,
    "b": Array [
      1,
    ],
  },
  "html": "
<div>
  <p>a</p>
  <p>b</p>
</div>
",
}
`;

exports[`should place placeholders for multiply "if" and "for" tags 1`] = `
Object {
  "Options": Object {
    "ID": "component",
  },
  "html": "
<div>
  <!--0-->
  <!--1-->
</div>
",
}
`;

exports[`should place placeholders for multiply "if" tags 1`] = `
Object {
  "Options": Object {
    "ID": "component",
  },
  "html": "
<div>
  <!--0-->
  <!--1-->
</div>
",
}
`;

exports[`should properly for with filters 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "filters": Object {
      "append": [Function],
      "upperCase": [Function],
    },
    "text": "upper_",
  },
  "html": "<p>UPPER_CASE</p>",
}
`;

exports[`should properly work with more then one node on topmost level 1`] = `
Object {
  "Options": Object {
    "ID": "component",
  },
  "html": "
<p>first</p>
<p>second</p>
",
}
`;

exports[`should properly work with text constants in attributes 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "bar": "bar",
  },
  "html": "<div class=\\"foo bar baz\\"></div>",
}
`;

exports[`should properly work with text constants in text nodes 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "bar": "bar",
  },
  "html": "
<p>foo
  <!--0-->bar
  <!--1--> baz
</p>
",
}
`;

exports[`should render attributes without quotes 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "name": "name",
  },
  "html": "<div class=\\"name\\"></div>",
}
`;

exports[`should render empty attributes 1`] = `
Object {
  "Options": Object {
    "ID": "component",
  },
  "html": "<input type=\\"checkbox\\" value=\\"\\" checked=\\"\\">",
}
`;

exports[`should render simple DOM 1`] = `
Object {
  "Options": Object {
    "ID": "component",
  },
  "html": "<div>Content</div>",
}
`;

exports[`should replace HTML entities with Unicode symbols 1`] = `
Object {
  "Options": Object {
    "ID": "component",
  },
  "html": "\\"&amp;'&lt;&gt;©£±¶ — € ♥¬Exists;",
}
`;

exports[`should save value for variables in complex cases 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "bar": "second",
    "foo": "first",
  },
  "html": "<div class=\\"first second\\"></div>",
}
`;

exports[`should save value for variables in complex cases 2`] = `
Object {
  "Options": Object {
    "ID": "component",
    "bar": "second",
    "foo": "updated",
  },
  "html": "<div class=\\"updated second\\"></div>",
}
`;

exports[`should support don't render checked == false attr 1`] = `
Object {
  "Options": Object {
    "ID": "component",
  },
  "html": "<input type=\\"checkbox\\">",
}
`;

exports[`should support expressions without variables 1`] = `
Object {
  "Options": Object {
    "ID": "component",
  },
  "html": "7",
}
`;

exports[`should support global variables 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "host": "sham-ui-ssr.example.com",
  },
  "html": "
<i>expr
  <!--0-->,
  <!--1-->if
  <!--0-->,
  <!--2-->for
  <!--1-->, <i class=\\"attr\\"></i>
</i>
",
}
`;

exports[`should support global variables 2`] = `
Object {
  "Options": Object {
    "ID": "component",
    "array": Array [
      1,
      2,
      3,
    ],
    "obj": Object {
      "a": 1,
      "b": 2,
    },
  },
  "html": "
array
<!--0-->,
<!--1-->4
<!--2-->,
<!--3-->a;b
<!--4-->,
<!--5-->{\\"a\\":1,\\"b\\":2}
",
}
`;

exports[`should support render checked == true attr 1`] = `
Object {
  "Options": Object {
    "ID": "component",
  },
  "html": "<input type=\\"checkbox\\" checked=\\"\\">",
}
`;

exports[`should support style attribute 1`] = `
Object {
  "Options": Object {
    "ID": "component",
  },
  "html": "<div style=\\"background-color: red\\">DANGER</div>",
}
`;
